### ISSUE TITLE
[Discussion] Add custom device

### ISSUE THREAD

**Issue by accelerate321** [pytorch/pytorch#83657]
### 🚀 The feature, motivation and pitch

We are an accelerator vendor and provide Pytorch extensions for our hardware MLU(https://github.com/Cambricon/catch). We would like to add a custom dispatch key and notice that other vendors have submitted their requests like XPU, IPU or NPU(only parts of dispatch keys are accepted), and you recommend to use PrivateUse key instead. We know the limits of adding a new key in Pytorch files but still have some questions.

1. Are all following vendors unable to merge their own custom dispatch key?

2. In some case, PrivateUse key can not meet demands. We think each kind of device may have some special cases to handle like:

https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/TensorShape.cpp#L1305
```
if (!self.is_xla() && !self.is_lazy() && !self.is_ipu()) { 
    return self._reshape_alias(shape, stride.value());
} else {
    return self.view(shape);
}
```

https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/TensorIterator.cpp#L1497
```
if (common_device_.type() == DeviceType::XLA  ||
      common_device_.type() == DeviceType::IPU  ||
      common_device_.type() == DeviceType::Lazy ||
      common_device_.type() == DeviceType::ORT  ||
      common_device_.type() == DeviceType::HPU) return;
```
It is easy to deal with such conditional judgement if our custom key is merged, but how to solve the problem when we do not change Pytorch files and only use PrivateUse key?

**Comment by gchanan** [comment]
@bdhirsh

**Comment by bdhirsh** [comment]
@accelerate321 

I think that long-term, usage of `PrivateUse` keys for external backends is the way we want to go. Any examples of hardcoding specific backends like ORT/HPU into parts of the codebase are not the direction we want to go in, and we should consider removing them over time.

As for the the two examples that you listed, and how they should be handled as a `PrivateUse` external backend:

> https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/TensorIterator.cpp#L1497

I think we probably want to add `PrivateUse1` check in that condition too. The main use case is that TensorIterator has a lot of complicated logic for shape inference, and it's technically possible for backends to try to re-use that logic. If you aren't actively attempting to use TensorIterator to perform shape inference for your own kernels then this shouldn't be a problem.

> https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/TensorShape.cpp#L1305

This check is a tiny performance optimization for `reshape()`: for strided backends (like CPU/CUDA), we already calculated the strides above, so rather than calling `.view()` (which will recompute the strides a second time), we call another opt that takes in the strides directly. This isn't important for non-strided backends though, like XLA/LazyTensor.

By default today, it looks like open registration backends (using `PrivateUse1`) will call the fast path logic `_reshape_alias()`, and assume that your backend supports strides. However, with `PrivateUse1` you have the power to override that. If that decomposition for reshape doesn't work with your backend, you can write a new one, and register it directly to the `AutogradPrivateUse1` and `PrivateUse1` dispatch keys. That will force you custom decomposition for reshape to run, instead of the default one in core.

Hope that answers your questions!

Separately, there's also now a slightly more full-featured example of using open registration for a custom backend here: https://github.com/bdhirsh/pytorch_open_registration_example (cc @albanD)

**Comment by accelerate321** [comment]
Thanks for your reply! If we want to reuse some logic related to specific device like TensorIterator above, do you mean apply patches about privateuse key on Pytorch? 
Actually we don't want to add any code or apply any patch to Pytorch. For this aim, it seems that we have to write similar code in our own repository and cannot reuse your logic. Is what I say right? Or is there any other suggestion?

**Comment by bdhirsh** [comment]
> If we want to reuse some logic related to specific device like TensorIterator above, do you mean apply patches about privateuse key on Pytorch?

Yep - I think we'll want to patch that in if you want to re-use TensorIterator code in your backend.

> Actually we don't want to add any code or apply any patch to Pytorch. For this aim, it seems that we have to write similar code in our own repository and cannot reuse your logic. Is what I say right? Or is there any other suggestion?

That specific patch for TensorIterator should be easy :) (TensorIterator is able to handle a lot of edge cases by handling type promotion + broadcasting for you, when you need to figure out what the shape of the output tensor should be). But yep - there's no requirement that you need to re-use TensorIterator from core if you're writing your own kernels
